### ISSUE TITLE
Universal MelGAN

### ISSUE THREAD

**Issue by huseinzol05** [TensorSpeech/TensorFlowTTS#506]
Original paper from https://arxiv.org/abs/2011.09631

Thanks to this repository, I added [STFTDiscriminator](https://github.com/huseinzol05/malaya-speech/blob/master/malaya_speech/train/model/universal_melgan/model.py#L328) and follow parameters from the paper, https://github.com/huseinzol05/malaya-speech/blob/master/malaya_speech/config/universal_melgan.py

I able to release pretty good Universal MelGAN and smaller factor Universal MelGAN with 1024 filters, can check examples at https://malaya-speech.readthedocs.io/en/latest/load-universal-melgan.html

### Code
1. Model definition, https://github.com/huseinzol05/malaya-speech/blob/master/malaya_speech/train/model/universal_melgan/model.py
2. Config definition, https://github.com/huseinzol05/malaya-speech/blob/master/malaya_speech/config/universal_melgan.py
3. Training definition, https://github.com/huseinzol05/malaya-speech/tree/master/pretrained-model/vocoder/universal-melgan

### Preprocess steps
1. Change into melspectrogram.
2. log 10 that melspectrogram.
3. Normalize using global mean and std.

For  no 3, no need to do global norm, scale to -1 and 1 is good enough, tested in small dataset.

**Comment by dathudeptrai** [comment]
@huseinzol05 great!. Can we somehow support Ur universal pretrained for this repo ?

**Comment by huseinzol05** [comment]
But Text-to-Mel models in this repository got their own norm, need to un-norm and norm using my Universal MelGAN.

**Comment by dathudeptrai** [comment]
@huseinzol05 can you point out where is a diff between universal melgan / melgan in terms of modeling?

**Comment by huseinzol05** [comment]
1. STFTDiscriminator, https://github.com/huseinzol05/malaya-speech/blob/master/malaya_speech/train/model/universal_melgan/model.py#L343
2. STFTDiscriminator required TFReflectionPad2d, https://github.com/huseinzol05/malaya-speech/blob/master/malaya_speech/train/model/universal_melgan/model.py#L71
3. MultiScaleDiscriminator got both WaveFormDiscriminator and STFTDiscriminator, https://github.com/huseinzol05/malaya-speech/blob/master/malaya_speech/train/model/universal_melgan/model.py#L486
4. Config for Universal MelGAN https://github.com/huseinzol05/malaya-speech/blob/master/malaya_speech/config/universal_melgan.py

### Preprocess steps

1. Change into melspectrogram.
2. log 10 that melspectrogram.
3. Norm using global norm / z-score, or scale to any boundary you want. My models use global norm, but I tested scaling between -1 and 1, also worked.

### Training session

Training session same as normal MelGAN training session. I trained on VCTK and malay speakers, feel free to download malay speakers dataset at https://github.com/huseinzol05/malaya-speech/tree/master/data/semisupervised-audiobook.

Tested on unseen malay, english and hindi speakers, really really good, will update the test set later.

**Comment by OnceJune** [comment]
how's the rtf of 1024 filters?

**Comment by huseinzol05** [comment]
what is rtf? lol, sorry, I got many definitions here

**Comment by dathudeptrai** [comment]
> what is rtf? lol, sorry, I got many definitions here

real time factor :D

**Comment by huseinzol05** [comment]
https://gist.github.com/huseinzol05/4d0d813adf7c35de9f025a78d34fd3f1

RTF:
1. single speaker melgan (17.3MB): 0.07561842918375423
2. universal melgan (309MB): 1.1712138246206785
3. universal melgan 1024 (78.4MB): 0.3548741817549521

**Comment by ZDisket** [comment]
Have you tried universal multiband melgan?

**Comment by huseinzol05** [comment]
I believe it can!
Let me try, i will post results in 1-3 weeks.

**Comment by wblgers** [comment]
Is this familiar with TFGAN?  Link below: https://arxiv.org/abs/2011.12206
